### ISSUE TITLE
Add support for AF_UNIX to IReactorSocket.adoptStreamPort

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5573]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5573|
|Type|enhancement|
|Created|2012-03-27 14:11:45Z|
|Branch|https://github.com/twisted/twisted/tree/unix-adopt-port-5573-3|

[#5248](https://github.com/twisted/twisted/issues/5248) adds support for re-using an existing AF_INET or AF_INET6 socket as a listening port.  `adoptStreamPort` should support AF_UNIX as well (and doing so should be simple).

Attachments:

* [5573-adopt-unix-sockets-v1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b8b/b8bc129977d38d110096beea8d3f79f6e381b6ea/a7546260fd3562005ee42904cfd79c986e4ee5fd.patch) (11095 bytes) - added by getoffmalawn on 2015-02-28 06:54:57Z - 
* [5573-adopt-unix-sockets-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b8b/b8bc129977d38d110096beea8d3f79f6e381b6ea/e91b4caa8ca35ce713f864696df618f839ddb7b8.patch) (9079 bytes) - added by getoffmalawn on 2015-02-28 09:56:42Z - Patch without exception style changes
* [5573-adopt-unix-sockets-v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b8b/b8bc129977d38d110096beea8d3f79f6e381b6ea/fada828af64fb07ff566b5ffdc67b19e1915b1b3.patch) (15170 bytes) - added by getoffmalawn on 2015-03-01 10:09:40Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5573 5573
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_unix_adopt_port_5573_3 branches/unix-adopt-port-5573-3
branch_author__glyph__getoffmalawn glyph, getoffmalawn
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1332857505000000 1332857505000000
changetime__1497589942664036 1497589942664036
version__None None
owner__glyph glyph
cc__glyph cc__morgen cc__cwillu cc__getoffmalawn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This will be easier after [#5578](https://github.com/twisted/twisted/issues/5578).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Once this is implemented, the systemd endpoint should work with UNIX sockets.  If it does, then the systemd documentation which mentions that UNIX sockets are not supported should be adjusted.  Otherwise, re-open [#6310](https://github.com/twisted/twisted/issues/6310) to make it work.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [37573]) Branching to 'unix-adopt-port-5573'

**Comment by twisted-trac** [comment]
|<img alt="getoffmalawn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|getoffmalawn set owner to getoffmalawn<br>getoffmalawn set status to `assigned`|
|-|-|

I've attached a patch that gets UNIX socket adoption working for adoptStreamPort and adoptStreamConnection. I've had some trouble getting adoptDatagramPort working, but if I get that going I'll attach a patch to [#6954](https://github.com/twisted/twisted/issues/6954).

A couple of notes about the patch:

 - I've changed all the old-style exception raising to new-style. I'm happy to remove that to minimise changes, if desirable.
 - I haven't got any tests for adoptStreamConnection, and I couldn't find any. In my eyes I basically need to start a server outside of Twisted's vision, accept a connection, and then adopt that connection. If anyone has any advice on writing a test like this or has any similar tests, I'd be happy to have another go at it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [getoffmalawn](#note_7):

>  - I've changed all the old-style exception raising to new-style. I'm happy to remove that to minimise changes, if desirable.

Yes, please do that.  Smaller changes make it through review faster.  Those changes will be great, just please put them into a different patch and ticket :).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="getoffmalawn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|getoffmalawn commented|
|-|-|

Replying to [glyph](#note_9):
> Yes, please do that.  Smaller changes make it through review faster.  Those changes will be great, just please put them into a different patch and ticket :).
> 
> Thanks!

Sure! No problem, I've attached that patch now. I have a few other syntactic changes to make to twisted.internet.unix to make it Python 3 compatible, so I might bundle them up with those and submit it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to getoffmalawn|
|-|-|

> I haven't got any tests for adoptStreamConnection, and I couldn't find any. In my eyes I basically need to start a server outside of Twisted's vision, accept a connection, and then adopt that connection. If anyone has any advice on writing a test like this or has any similar tests, I'd be happy to have another go at it.

There are various existing tests for this

```
exarkun@top:~/Projects/Twisted/trunk$ grep adoptStreamConnection twisted/ -r --include 'test_*.py'
twisted/internet/test/test_tcp.py:    Test server transports built using C{adoptStreamConnection}.
twisted/internet/test/test_tcp.py:        a C{adoptStreamConnection} against the original server connection.
twisted/internet/test/test_tcp.py:            reactor.adoptStreamConnection(
twisted/internet/test/test_socket.py:    Builder for testing L{IReactorSocket.adoptStreamConnection}
twisted/internet/test/test_socket.py:        An implementation of L{IReactorSocket.adoptStreamConnection} raises
twisted/internet/test/test_socket.py:            reactor.adoptStreamConnection, connection.fileno(), arbitrary,
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Also, you can look at the revision that introduced that functionality: https://twistedmatrix.com/trac/changeset/34031 (note that trac doesn't rendered the _added_ files on that page, one of which a new test module, though).

**Comment by twisted-trac** [comment]
|<img alt="getoffmalawn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|getoffmalawn commented|
|-|-|

Replying to [exarkun](#note_12):
> 
> There are various existing tests for this
> 
> ```
> exarkun@top:~/Projects/Twisted/trunk$ grep adoptStreamConnection twisted/ -r --include 'test_*.py'
> twisted/internet/test/test_tcp.py:    Test server transports built using C{adoptStreamConnection}.
> twisted/internet/test/test_tcp.py:        a C{adoptStreamConnection} against the original server connection.
> twisted/internet/test/test_tcp.py:            reactor.adoptStreamConnection(
> twisted/internet/test/test_socket.py:    Builder for testing L{IReactorSocket.adoptStreamConnection}
> twisted/internet/test/test_socket.py:        An implementation of L{IReactorSocket.adoptStreamConnection} raises
> twisted/internet/test/test_socket.py:            reactor.adoptStreamConnection, connection.fileno(), arbitrary,
> ```

Great! Thank you. I've added tests now and submitted a new patch. There is one note that I forgot to mention earlier, in `twisted.internet.unix.Server._fromConnectedSocket()`, I wasn't sure of the best place to draw a session number from, so I've set it to zero:

```
    # FIXME: is this a suitable sessionno?
    sessionno = 0
    self = cls(skt, proto, skt.getpeername(), None, sessionno, reactor)
```

In `twisted.internet.tcp.Server._fromConnectedSocket()`, the session number is set to the port of the connected peer - given the UNIX socket doesn't have an equivalent, I'm not sure what to do here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

(In [44046]) Branching to unix-adopt-port-5573-3.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Patch uploaded and branched kicked off for the next reviewer.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban set owner to getoffmalawn|
|-|-|

Thanks for the patch.

Changes looks good but some tests are failing and they need to be fixed before landing.
I guess that Unix datagram test should be skipped on Windows.

Please note that I am still a junior reviewer so I might suggest stupid things. Feel free to ignore them.

-----

I am not very happy with

```
assert unixEnabled, "UNIX support is not present"
```

In case we want raise I prefer the explicit method

```
if not unixEnabled:
   raise NotImplementedError('UNIX support is not present')
```

and write a test to touch this code and check that assertion is raised.

BUT, he code already raises UnsupportedAddressFamily so I think that we should just remove this assert.


-----

All new methods/functions need documentation for all arguments and return value. ex getConnectedClientAndServer and getListeningPort

Due to missing documentation, twistedchecker is failing.

----

For FIXME/TODO comments we should create a dedicated ticket and add the URL to the ticket in that comment.

For 

```
sessionno = 0
```

I think that is ok to open a ticket but also we need to document this behaviour.
Maybe in `Limitations and Known Issues` we can mention that for unix sockets sessionno is always 0

-----

self._preexistingSocket needs to be documented in class docstring even if it is private. In this way other developers should be aware of it and prevent an accidental clash with another method.

what is self.repstr ?

-----

I think that newsfile should be a single line as https://github.com/twisted/newsbuilder will rewrap newlines.
Also, it should be a .feature and not a .misc as content for .misc is ignored.

.misc tickets are only listed in the also fixed section with only ticket ID mentioned.

You might want to reformulate it as

```
twisted.internet.posixbase.PosixReactorBase.adoptStreamPort and twisted.internet.posixbase.PosixReactorBase.adoptStreamConnection now supports UNIX SOCK_STREAM socket types.
```

------

in test_ServerAddressUNIX
```
Helper method to test TCP server addresses on either IPv4 or IPv6
```

I think that this should be UNIX server not TCP
-----


For

```
    def getListeningPort(self, reactor, factory):
        """
        Get a UNIX port from a reactor, wrapping an already-initialized file
        descriptor.
        """
        if IReactorSocket.providedBy(reactor):
            portSock = socket(AF_UNIX)
            # Code here
        else:
            raise SkipTest("Reactor does not provide IReactorSocket")
```

maybe we can have it as

```
    def getListeningPort(self, reactor, factory):
        """
        Get a UNIX port from a reactor, wrapping an already-initialized file
        descriptor.
        """
        if not IReactorSocket.providedBy(reactor):
            raise SkipTest("Reactor does not provide IReactorSocket")

        portSock = socket(AF_UNIX)
        # More code here
            
```

------

Instead of

```
            # self.mktemp() often returns a path which is too long to be used.
            path = mktemp(suffix='.sock', dir='.')
```

maybe we can have


```
path = self.getSocketPath()

def getUnixSocketPath(self):
    """
    Return a path which can be used as an Unix socket.
    """
    # self.mktemp() often returns a path which is too long to be used.
    path = mktemp(suffix='.sock', dir='.')
```

In this way we don't have to duplicate the comment and in case we find out a better way to generate unix socket path we only have to change the code in one place.

-------

I don't know what to say about this

```
[peerAddress] = server.factory.peerAddresses
```

it works but in the other Twisted code I see that it uses

```
peerAddress = server.factory.peerAddresses[0]
``` 


--------

Regarding test_ServerAddressUNIX I think that  we should do the assertion synchronous.

```
protocols = []

d.addCallback(lambda result: protocols.append(protocols))
self.runReactor(reactor)

self.assertEqual([(some, things)], protocols)

```

In this way we make sure that callback is called.
Without this the callback might not be called but test will still pass.

-------

Do we want log.msg in tests?

If yes, then please add a comment why we need this and fix the string formating. ie
instead of
```
"string %s" % value
```
use
```
"string %s % (value,)
```

-----


So this patch still needs changes before being merged.

Please check my comments!

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exvito's avatar" src="https://avatars.githubusercontent.com/u/7661068?s=50" width="50" height="50">](https://github.com/exvito)<a name="note_19"></a>|@exvito set owner to @exvito|
|-|-|

All,

I'm volunteering to pick this up where it was left off. Will work from latest patch + review onwards.

Input is welcome.

**Comment by twisted-trac** [comment]
|[<img alt="exvito's avatar" src="https://avatars.githubusercontent.com/u/7661068?s=50" width="50" height="50">](https://github.com/exvito)<a name="note_20"></a>|@exvito commented|
|-|-|

PROGRESS:
* Merged in latest patch.
* Addressed feedback from Adi Roiban.
* Added minor fixes to make tests pass on Python 2 and 3 on Linux and Mac OS.
* Submited PR at https://github.com/twisted/twisted/pull/742
* Waiting to see test results.
* Will request review if tests pass.

**Comment by twisted-trac** [comment]
|[<img alt="exvito's avatar" src="https://avatars.githubusercontent.com/u/7661068?s=50" width="50" height="50">](https://github.com/exvito)<a name="note_21"></a>|@exvito removed owner|
|-|-|

Reviewers,

This has been on the backburner for a while. I've revisited today and while there's still an oddity with github's codecov/patch check that I can't seem to understand fully, other checks are passing and the code is ready for review.

In summary, I picked up a 2 year old patch and adjusted slightly per the previous reviews here and, of course, to ensure tests pass. There's more info and context in the PR at https://github.com/twisted/twisted/pull/742.

Thanks in advance.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set owner to @glyph|
|-|-|

L G T M.  I especially appreciate your attention to detail updating the documentation.  Thank you for picking up this old patch!

Will merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph set status to `closed`|
|-|-|

In changeset 66db5c752f90639f8a43720dce584682f7e0e387
```
#!CommitTicketReference repository="" revision="66db5c752f90639f8a43720dce584682f7e0e387"
Merge pull request #742 from exvito/5573-exvito-unix-adopt-stream

Author: exvito

Reviewer: glyph

Fixes: ticket:5573

twisted.internet.posixbase.PosixReactorBase.adoptStreamPort and twisted.internet.posixbase.PosixReactorBase.adoptStreamConnection now support AF_UNIX SOCK_STREAM sockets
```
